skbio/alignment/_path.py

```diff
@@ -0,0 +1,447 @@
+        idx = np.append(0, np.where(ints[:-1] != ints[1:])[0] + 1)
+        lens = np.append(idx[1:] - idx[:-1], ints.size - idx[-1])
+        return cls(lens, ints[idx], np.zeros(bits.shape[0], dtype=int))
+
+    def to_bits(self):
+        r"""Unpack states into an array of bits."""
+        if not np.all(np.isin(self.states, [0, 1, 2, 3])):
+            raise ValueError(
+                "For pairwise alignment, `states` must only contain "
+                "zeros, ones, twos, or threes."
+            )
+        return np.stack([self.states & 1, self.states >> 1])
+
+    def to_cigar(self, seqs=None):
+        r"""Generate a CIGAR string representing the pairwise alignment path.
+
+        Parameters
+        ----------
+        seqs : list of skbio.Sequence or string
+            A pair of sequences to generate CIGAR string. If provided, will
+            distinguish match (``=``) and mismatch (``X``). Otherwise, will uniformly
+            note them as (mis)match (``M``). The first sequence in the list is the
+            query sequence, the second is the reference sequence.
+        """
+        cigar = []
+
+        states = np.squeeze(self.states)
+
+        if seqs is not None:
+            # test if seqs is strings or Sequence object or something else
+            if isinstance(seqs[0], str) and isinstance(seqs[1], str):
+                seq1 = np.frombuffer(seqs[0].encode("ascii"), dtype=np.uint8)
+                seq2 = np.frombuffer(seqs[1].encode("ascii"), dtype=np.uint8)
+            elif isinstance(seqs[0], Sequence) and isinstance(seqs[1], Sequence):
+                seq1 = seqs[0]._bytes
+                seq2 = seqs[1]._bytes
+            else:
+                raise ValueError("`seqs` must be strings or Sequence objects.")
+
+            idx1, idx2 = [int(x) for x in self.starts]
+
+            for length, state in zip(self.lengths, states):
+                if state == 0:
+                    match_arr = seq1[idx1 : idx1 + length] == seq2[idx2 : idx2 + length]
+                    char_arr = np.where(match_arr, "=", "X")
+                    n = len(char_arr)
+                    count = 1
+                    curr_char = char_arr[0]
+                    for i in range(1, n):
+                        if char_arr[i] == curr_char:
+                            count += 1
+                        else:
+                            cigar.append(str(count) + curr_char)
+                            curr_char = char_arr[i]
+                            count = 1
+                    cigar.append(str(count) + curr_char)
+                    idx1 += length
+                    idx2 += length
+                elif state == 1:
+                    cigar.append(str(length) + "I")
+                    idx2 += length
+                elif state == 2:
+                    cigar.append(str(length) + "D")
+                    idx1 += length
+                elif state == 3:
+                    cigar.append(str(length) + "P")
```

Comment:
There is a shorter may that can replace the code between lines 332 to 354:

```python
if state == 0:
    match_arr = ...
else:
    cigar.append(str(length) + codes[state])
flipped = state ^ 3
idx1 += flipped & 1 and length
idx2 += flipped & 2 and length
```

This method uses indexing and bitwise operations to avoid rewriting `cigar.append` and `idx +=` for each condition. It is not necessarily faster than the current solution, nor is it more elegant (bitwise operation is hard to understand!). Therefore, it is only for reference.